test/unity/toolkits/drawing_classifier/test_dc_prediction.cxx

```diff
@@ -0,0 +1,584 @@
+
+/**
+ * tuple of batch_size, num_of_rows, num_of_classes
+ *
+ * batch_size, num_of_rows, num_of_classes
+ * 1. batch_size > num_of_rows
+ * 2. batch_size == num_of_rows
+ * 3. num_of_rows can divide batch_size
+ * 4. num_of_rows can not divide batch_size
+ *
+ **/
+std::vector<std::tuple<unsigned, unsigned, unsigned>> TEST_CASES = {
+    {2, 1, 1}, {2, 1, 2}, {2, 2, 2}, {2, 4, 2},
+    {2, 4, 3}, {2, 5, 2}, {2, 5, 3}};
+
+constexpr auto PRED_NAME = "preds";
+
+/* ============== utils borrowed from gl_sframe.cxx ============= */
+// TODO: create a common util class
+void _assert_flexvec_equals(const std::vector<flexible_type>& sa,
+                            const std::vector<flexible_type>& sb) {
+  TS_ASSERT_EQUALS(sa.size(), sb.size());
+  for (size_t i = 0; i < sa.size(); ++i) {
+    if (sa[i] != sb[i]) {
+      std::cout << sa[i] << " vs " << sb[i] << std::endl;
+    }
+    TS_ASSERT_EQUALS(sa[i], sb[i]);
+  }
+}
+
+void _assert_sframe_equals(gl_sframe sa, gl_sframe sb) {
+  TS_ASSERT_EQUALS(sa.size(), sb.size());
+  TS_ASSERT_EQUALS(sa.num_columns(), sb.num_columns());
+  auto a_cols = sa.column_names();
+  auto b_cols = sb.column_names();
+  std::sort(a_cols.begin(), a_cols.end());
+  std::sort(b_cols.begin(), b_cols.end());
+  for (size_t i = 0; i < a_cols.size(); ++i)
+    TS_ASSERT_EQUALS(a_cols[i], b_cols[i]);
+  sb = sb[sa.column_names()];
+  for (size_t i = 0; i < sa.size(); ++i) {
+    _assert_flexvec_equals(sa[i], sb[i]);
+  }
+}
+
+/* ========================= test drivers & utils ======================== */
+
+#ifndef NDEBUG
+#define log_for_debug(x) \
+  { logprogress_stream << ">>>" << (x) << std::endl; }
+#else
+#define log_for_debug(x) {}
+#endif
+
+struct mock_perform_inference : public test_drawing_classifier {
+  inline gl_sframe get_inference_result(data_iterator* data) {
+    return perform_inference(data);
+  }
+};
+
+gl_sframe set_up_perform_inference(
+    test_drawing_classifier& mock_model,
+    std::unique_ptr<mock_model_backend>& mock_backend,
+    std::unique_ptr<mock_compute_context>& mock_context, size_t batch_size,
+    const size_t num_of_rows, const size_t num_of_classes) {
+  ASSERT_MSG(num_of_rows > 0, "num_of_rows should be bigger than 0");
+  ASSERT_MSG(num_of_classes > 0, "num_of_classes should be bigger than 0");
+  ASSERT_MSG(batch_size > 0, "batch_size should be bigger than 0");
+
+  gl_sframe expected_sf;
+  {
+    std::vector<float> buffer(num_of_classes * batch_size);
+
+    std::default_random_engine eng;
+    std::uniform_int_distribution<float> distribution(0, 20);
+
+    gl_sframe_writer writer({PRED_NAME}, {flex_type_enum::VECTOR},
+                            /* num_segments */ 1);
+
+    size_t ii = 0;
+    while (ii < num_of_rows) {
+      buffer.clear();
+
+      size_t repeat =
+          ii + batch_size > num_of_rows ? num_of_rows - ii : batch_size;
+
+      ii += batch_size;
+
+      for (size_t jj = 0; jj < repeat; ++jj) {
+        auto pbeg = buffer.begin() + jj * num_of_classes;
```

Comment:
Can we have better variable names here, maybe `begin_ptr` instead of `pbeg`, `end_ptr` instead of `pend`?

Reply:
Nah. pbeg or pend is tradition c style. It's widely used.